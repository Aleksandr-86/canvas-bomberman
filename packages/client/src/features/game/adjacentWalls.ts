import { EMPTY_FIELD, CELL_WIDTH } from './const'
import { Kind } from './types'
import { Point, type PointLike } from './utils'

/**
 * Gets cell value from point
 * @param p point (in game space)
 */
const getFieldCell = ({ x, y }: PointLike) => {
  return EMPTY_FIELD[y * CELL_WIDTH + x]
}

const unit = {
  left: new Point(-1, 0),
  right: new Point(1, 0),
  up: new Point(0, -1),
  down: new Point(0, 1),
}

/**
 * Create list of coordinates (screen space) of adjacent walls affected by explosion
 * @param walls list of coordinates to match (screen space)
 * @param center coordinate of explosion center (screen space)
 * @param radius explosion radius
 * @returns
 */
export function adjacentWalls(center: PointLike, radius: number) {
  const hits = []

  const cutoffs = {
    left: false,
    right: false,
    up: false,
    down: false,
  }

  const centerCell = new Point(center.x, center.y).apply(
    c => Math.floor(c / CELL_WIDTH) * CELL_WIDTH
  )

  hits.push(center)

  for (let i = 1; i <= radius; ++i) {
    const cellLeft = new Point(
      centerCell.x / CELL_WIDTH + -1 * i,
      centerCell.y / CELL_WIDTH
    )

    if (!cutoffs.left && getFieldCell(cellLeft) !== Kind.WallHard) {
      hits.push(cellLeft.scale(CELL_WIDTH))
    } else {
      cutoffs.left = true
    }

    const cellRight = new Point(
      centerCell.x / CELL_WIDTH + 1 * i,
      centerCell.y / CELL_WIDTH
    )
    if (!cutoffs.right && getFieldCell(cellRight) !== Kind.WallHard) {
      hits.push(cellRight.scale(CELL_WIDTH))
    } else {
      cutoffs.right = true
    }

    const cellUp = new Point(
      centerCell.x / CELL_WIDTH,
      centerCell.y / CELL_WIDTH + -1 * i
    )
    if (!cutoffs.up && getFieldCell(cellUp) !== Kind.WallHard) {
      hits.push(cellUp.scale(CELL_WIDTH))
    } else {
      cutoffs.up = true
    }

    const cellDown = new Point(
      centerCell.x / CELL_WIDTH,
      centerCell.y / CELL_WIDTH + 1 * i
    )
    if (!cutoffs.down && getFieldCell(cellDown) !== Kind.WallHard) {
      hits.push(cellDown.scale(CELL_WIDTH))
    } else {
      cutoffs.down = true
    }
  }

  return hits
}
