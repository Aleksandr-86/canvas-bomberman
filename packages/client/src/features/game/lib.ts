class Vec2 {
  static get Down() {
    return new Vec2(0, -1)
  }
  static get Up() {
    return new Vec2(0, 1)
  }
  static get Left() {
    return new Vec2(-1, 0)
  }
  static get Right() {
    return new Vec2(1, 0)
  }
  static get Zero() {
    return new Vec2(0, 0)
  }

  static isVec(something: unknown) {
    return something instanceof Vec2
  }

  constructor(public x: number, public y: number) {}
}

class Rect {
  private origin: Vec2 | 0 = 0
  private dimensions: Vec2 | 0 = 0

  constructor(position: Vec2, size: Vec2)
  constructor(x: number, y: number, width: number, height: number)
  constructor(
    positionOrX: number | Vec2,
    dimensionsOrY: number | Vec2,
    width?: number,
    height?: number
  ) {
    if (positionOrX instanceof Vec2 && dimensionsOrY instanceof Vec2) {
      this.origin = positionOrX
      this.dimensions = dimensionsOrY
    }

    if (
      width &&
      height &&
      typeof positionOrX === 'number' &&
      typeof dimensionsOrY === 'number'
    ) {
      this.origin = new Vec2(positionOrX, dimensionsOrY)
      this.dimensions = new Vec2(width, height)
    }
  }
}

export { Vec2, Rect }
