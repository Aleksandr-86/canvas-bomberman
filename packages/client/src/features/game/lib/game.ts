import { startGame } from '../gameActions'
import { Keyboard } from './keyboard'
import { Scene, type SceneContext } from './scene'
import { Ticker } from './ticker'

interface GameConfig {
  width: number
  heigth: number
  root: HTMLCanvasElement
  create: (scene: Scene) => void
  update: (ctx: SceneContext) => void
  backgroundColor: string
}

export class Game {
  private kbd: Keyboard = new Keyboard()
  private scene = new Scene()
  private ticker = new Ticker()
  private root: HTMLCanvasElement
  private ctx: CanvasRenderingContext2D
  public screenWidth: number
  public screenHeight: number
  private update: (ctx: SceneContext) => void
  private create: (scene: Scene) => void
  public started = false

  constructor({
    root,
    width,
    heigth,
    update,
    create,
    backgroundColor,
  }: GameConfig) {
    this.root = root
    this.screenWidth = width
    this.screenHeight = heigth
    this.update = update
    this.create = create

    const ctx = root.getContext('2d')

    if (!ctx) {
      throw new Error('Error getting context')
    }
    ctx.imageSmoothingEnabled = false

    this.ctx = ctx
    this.root.width = width
    this.root.height = heigth
    this.root.style.backgroundColor = backgroundColor
  }

  start() {
    this.create(this.scene)
    this.ticker.add(time => {
      this.update({
        time,
        kbd: this.kbd,
        scene: this.scene,
      })
      this.ctx.clearRect(0, 0, this.screenWidth, this.screenHeight)
      this.scene.render(this.ctx, time)
    })

    this.ticker.start()
    this.started = true
    startGame()
  }

  stop() {
    this.ticker.stop()
    this.started = false
  }
}
