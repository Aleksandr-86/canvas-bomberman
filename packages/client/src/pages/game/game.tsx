import { GameDisplay } from '../../components/gameDisplay/gameDisplay'

export const Game = () => {
  return (
    <>
      <h2>Game page</h2>
      <GameDisplay />
    </>
  )
}
