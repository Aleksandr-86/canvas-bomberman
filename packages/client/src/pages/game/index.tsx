import { GameScreen } from '../../widgets'

const Game = () => {
  return (
    <>
      <h2>Game page</h2>
      <GameScreen />
    </>
  )
}

export { Game }
