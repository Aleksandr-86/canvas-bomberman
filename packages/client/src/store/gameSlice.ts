import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import {
  PLAYER_STARTING_POSITION,
  EMPTY_FIELD,
  GRID_HEIGHT,
  GRID_WIDTH,
} from '../features/game/const'
import { Kind, type FieldState } from '../features/game/types'
import { delay, Vec2 } from '../features/game/utils'
import { fillSoftWalls } from '../features/game/fillSoftWalls'
import { explodeAdjacentSoftWalls } from '../features/game/explodeAdjacentSoftWalls'
import { AppDispatch } from './index'

export enum GameStatus {
  START = 'START',
  IN_PROGRESS = 'IN_PROGRESS',
  END = 'END',
}

type GameState = {
  field: FieldState
  status: GameStatus
  currentScore: number
  playerPosition: Vec2
  doorPosition: Vec2
  bombs: Vec2[]
  time: number
}

const INITIAL_STATE: GameState = {
  status: GameStatus.START,
  currentScore: 0,
  doorPosition: new Vec2(GRID_WIDTH - 2, GRID_HEIGHT - 2),
  playerPosition: PLAYER_STARTING_POSITION,
  bombs: [],
  field: EMPTY_FIELD,
  time: 300,
}

const gameSlice = createSlice({
  name: 'game',
  initialState: () => {
    return {
      ...INITIAL_STATE,
      field: fillSoftWalls(EMPTY_FIELD),
    }
  },
  reducers: {
    setStatus: (state, action: PayloadAction<GameStatus>) => {
      state.status = action.payload
    },
    setCurrentScore: (state, action: PayloadAction<number>) => {
      state.currentScore = action.payload
    },
    playerMoved: (state, { payload }: PayloadAction<Vec2>) => {
      state.playerPosition = payload
    },
    playerReset: state => {
      state.playerPosition = PLAYER_STARTING_POSITION
    },
    bombPlaced: (state, { payload: { x, y } }: PayloadAction<Vec2>) => {
      state.field[y][x] = Kind.Bomb
    },
    scoreIncreased: (state, { payload }: PayloadAction<number>) => {
      state.currentScore += payload
    },
    bombExploded: ({ field }, { payload }: PayloadAction<Vec2>) => {
      explodeAdjacentSoftWalls(field, payload)
    },
    timerUpdated: (state, { payload }: PayloadAction<number>) => {
      state.time -= payload * 0.001 /* payload is in ms */
    },
  },
})

export const bombSet = (pos: Vec2) => async (dispatch: AppDispatch) => {
  dispatch(bombPlaced(pos))
  await delay(800)
  dispatch(bombExploded(pos))
}

let previousTime = performance.now()
let gameIntervalId = 0

export const gameStarted = () => (dispatch: AppDispatch) => {
  gameIntervalId = window.setInterval(() => {
    const now = performance.now()
    const dt = now - previousTime
    previousTime = now
    dispatch(timerUpdated(dt))
  }, 1000)
  dispatch(setStatus(GameStatus.IN_PROGRESS))
}

export const gameEnded = () => (dispatch: AppDispatch) => {
  clearInterval(gameIntervalId)
  dispatch(setStatus(GameStatus.END))
}

export const gameReducer = gameSlice.reducer
export const {
  setStatus,
  setCurrentScore,
  playerMoved,
  playerReset,
  bombPlaced,
  bombExploded,
  scoreIncreased,
  timerUpdated,
} = gameSlice.actions
