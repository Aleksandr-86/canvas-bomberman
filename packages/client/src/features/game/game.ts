import { Vec2 } from './lib'

interface GameConfig {
  width: number
  heigth: number
  backgroundColor: string
  root: HTMLCanvasElement
}

interface Renderable {
  origin: Vec2
  dimensions: Vec2
}

class Game {
  private root: HTMLCanvasElement

  private scene: Renderable[] = []

  private ctx: CanvasRenderingContext2D

  constructor({ root, backgroundColor }: GameConfig) {
    this.root = root
    this.root.style.backgroundColor = backgroundColor
    const ctx = root.getContext('2d')
    if (!ctx) {
      throw new Error('Error getting canvas context')
    }
    this.ctx = ctx
  }

  public start() {
    const cb = () => {
      this.scene.forEach(({ origin, dimensions, color }) => {
        this.ctx.clearRect(0, 0, 10000, 100000)

        if (color) {
          this.ctx.fillStyle = color
        }

        this.ctx.fillRect(origin.x, origin.y, dimensions.x, dimensions.y)
        this.ctx.fillStyle = ''
      })
      window.requestAnimationFrame(cb)
    }
    cb()
  }

  public add(gameObject: Renderable) {
    this.scene.push(gameObject)
  }
}

class Container implements Renderable {
  constructor(
    public origin: Vec2,
    public dimensions: Vec2,
    public color: string = 'yellow'
  ) {}

  public move(direction: Vec2) {
    console.log('old dir', this.origin)
    this.origin = new Vec2(
      direction.x + this.origin.x,
      direction.y + this.origin.y
    )
    console.log('new dir', this.origin)
  }
}

export { Game, Container }
