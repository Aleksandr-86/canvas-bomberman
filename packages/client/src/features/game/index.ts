import { useRef, useEffect } from 'react'
import { useAppDispatch } from '../../store/hooks'
import { bombermanScene } from './bombermanScene'
import { setStatus, GameStatus } from '../../store/gameSlice'
import { Game } from '../game/lib'
import { GameConfig } from './lib'
import { gameEnded } from './gameActions'

export function createGame(config: GameConfig) {
  return new Game(config)
}

export function useGame() {
  const canvasRef = useRef<null | HTMLCanvasElement>(null)
  const gameRef = useRef<null | Game>(null)
  const shouldRunSecondTime = useRef(true)
  const dispatch = useAppDispatch()

  const endGame = () => {
    gameRef.current?.stop()
    gameRef.current = null
    gameEnded()
  }

  const startGame = () => {
    if (!canvasRef.current) {
      return
    }

    gameRef.current = createGame({
      height: 640,
      width: 1280,
      backgroundColor: '#64b0ff',
      root: canvasRef.current,
      scene: bombermanScene,
    })

    gameRef.current.start()
    dispatch(setStatus(GameStatus.START))
  }

  useEffect(() => {
    const canvas = canvasRef.current

    if (!canvas || !shouldRunSecondTime.current) {
      return
    }

    shouldRunSecondTime.current = false

    gameRef.current = new Game({
      height: 640,
      width: 1280,
      backgroundColor: '#64b0ff',
      root: canvas,
      scene: bombermanScene,
    })
  }, [])

  return {
    endGame,
    startGame,
    canvasRef,
    gameRef,
  }
}
