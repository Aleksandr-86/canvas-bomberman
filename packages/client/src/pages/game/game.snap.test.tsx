import * as renderer from 'react-test-renderer'
import { Game } from './game'
import configureStore from 'redux-mock-store'
import { GameStatus } from '../../store/game/gameSlice'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

const mockStore = configureStore([])

describe(`<Game/> snapshot test cases`, () => {
  test(`<Game/> should render <div className=gameDisplay> parent with <GameStart/> child component (status=START)`, () => {
    const store = mockStore({ game: { status: GameStatus.START } })

    const tree = renderer
      .create(
        <Provider store={store}>
          <Game />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test(`<Game/> should render <div className=gameDisplay> parent with child elements (status=IN_PROGRESS)`, () => {
    const store = mockStore({ game: { status: GameStatus.IN_PROGRESS } })

    const tree = renderer
      .create(
        <Provider store={store}>
          <Game />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test(`<Game/> should render <div className=gameDisplay> parent with <GameEnd/> child component (status=END)`, () => {
    const store = mockStore({ game: { status: GameStatus.END } })

    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter initialEntries={[`/game`]}>
            <Game />
          </MemoryRouter>
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
