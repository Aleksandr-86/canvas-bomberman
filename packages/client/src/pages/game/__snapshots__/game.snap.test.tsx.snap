// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> snapshot test cases <Game/> should render <div className=gameDisplay> parent with <GameEnd/> child component (status=END) 1`] = `
<div
  className="gameDisplay"
>
  <div
    className="gameDisplayCanvasWrapper"
  >
    <div
      className="panel"
    >
      <span>
        Time Left: 
        NaN
      </span>
      <span>
        Score: 
      </span>
    </div>
    <canvas
      height={640}
      width={1280}
    />
    <div
      className="gameOverlay"
    >
      <div
        className="gameEnd"
      >
        <p
          className="gameEndText"
        >
          Игра окончена!
        </p>
        <p
          className="gameEndText"
        >
          Количество набранных очков: 
        </p>
        <div
          className="gameEndButtons"
        >
          <a
            className="linkButton gameEndButton"
            href="/"
            onClick={[Function]}
          >
            На главную
          </a>
          <button
            className="button gameEndButton"
            onClick={[Function]}
          >
            Сыграть еще раз
          </button>
          <a
            className="linkButton gameEndButton"
            href="/leaderboard"
            onClick={[Function]}
          >
            Посмотреть таблицу лидеров
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Game/> snapshot test cases <Game/> should render <div className=gameDisplay> parent with <GameStart/> child component (status=START) 1`] = `
<div
  className="gameDisplay"
>
  <div
    className="gameDisplayCanvasWrapper"
  >
    <div
      className="panel"
    >
      <span>
        Time Left: 
        NaN
      </span>
      <span>
        Score: 
      </span>
    </div>
    <canvas
      height={640}
      width={1280}
    />
    <div
      className="gameOverlay"
    >
      <div
        className="gameStart"
      >
        <p>
          Куда жать, чтобы было весело?!
        </p>
        <img
          alt="Клавиатура"
          draggable={false}
          height={303}
          width={720}
        />
        <button
          className="button gameStartButton"
          onClick={[Function]}
        >
          ПОНЯЛ, ПОГНАЛИ!
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Game/> snapshot test cases <Game/> should render <div className=gameDisplay> parent with child elements (status=IN_PROGRESS) 1`] = `
<div
  className="gameDisplay"
>
  <div
    className="gameDisplayCanvasWrapper"
  >
    <div
      className="panel"
    >
      <span>
        Time Left: 
        NaN
      </span>
      <span>
        Score: 
      </span>
    </div>
    <canvas
      height={640}
      width={1280}
    />
  </div>
</div>
`;
